src/coap_client_zephyr.c

```diff
@@ -0,0 +1,1072 @@
+        goto free_req;
+    }
+
+    coap_req->block_ctx.current = (req->get_block.block_index * req->get_block.block_size);
+
+    err = golioth_coap_req_append_block2_option(coap_req);
+    if (err) {
+        LOG_ERR("Unable to append block2: %d", err);
+        goto free_req;
+    }
+
+    err = golioth_coap_req_schedule(coap_req);
+    if (err) {
+        LOG_ERR("Failed to schedule CoAP GET BLOCK: %d", err);
+        goto free_req;
+    }
+
+    return 0;
+
+free_req:
+    golioth_coap_req_free(coap_req);
+
+    return err;
+}
+
+static int golioth_coap_observe(golioth_coap_request_msg_t* req, struct golioth_client* client) {
+    int err;
+
+    err = golioth_coap_req_cb(req->client, COAP_METHOD_GET,
+                              PATHV(req->path_prefix, req->path),
+                              golioth_content_type_to_coap_format(req->observe.content_type),
+                              NULL, 0,
+                              golioth_coap_cb, req,
+                              GOLIOTH_COAP_REQ_OBSERVE);
+    if (err) {
+        LOG_ERR("Failed to schedule CoAP OBSERVE: %d", err);
+        return err;
+    }
+
+    return 0;
+}
+
+static void reestablish_observations(struct golioth_client* client) {
+    golioth_coap_observe_info_t* obs_info = NULL;
+    for (int i = 0; i < CONFIG_GOLIOTH_MAX_NUM_OBSERVATIONS; i++) {
+        obs_info = &client->observations[i];
+        if (obs_info->in_use) {
+            golioth_coap_observe(&obs_info->req, client);
+        }
+    }
+}
+
+static enum golioth_status coap_io_loop_once(struct golioth_client* client) {
+    golioth_coap_request_msg_t* req;
+    enum golioth_status status = GOLIOTH_OK;
+    int err = 0;
+
+    req = calloc(1, sizeof(*req));
+    if (!req) {
+        status = GOLIOTH_ERR_MEM_ALLOC;
+        goto free_req;
+    }
+
+    // Wait for request message, with timeout
+    bool got_request_msg = golioth_mbox_recv(
+        client->request_queue, req, CONFIG_GOLIOTH_COAP_REQUEST_QUEUE_TIMEOUT_MS);
+    if (!got_request_msg) {
+        // No requests, so process other pending IO (e.g. observations)
+        goto free_req;
+    }
+
+    // Make sure the request isn't too old
+    if (golioth_sys_now_ms() > req->ageout_ms) {
+        LOG_WRN("Ignoring request that has aged out, type %d, path %s",
+                req->type,
+                (req->path ? req->path : "N/A"));
+
+        if (req->type == GOLIOTH_COAP_REQUEST_POST && req->post.payload_size > 0) {
+            golioth_sys_free(req->post.payload);
+        }
+
+        if (req->request_complete_event) {
+            golioth_event_group_destroy(req->request_complete_event);
+            golioth_sys_sem_destroy(req->request_complete_ack_sem);
+        }
+
+        goto free_req;
+    }
+
+    req->client = client;
+
+    // Handle message and send request to server
+    switch (req->type) {
+        case GOLIOTH_COAP_REQUEST_EMPTY:
+            LOG_DBG("Handle EMPTY");
+            golioth_send_coap_empty(req->client);
```

Comment:
Should we capture the return value of this call?

Reply:
Assigned `err`, which should be then mapped to `status`. See https://github.com/golioth/golioth-firmware-sdk/compare/00f2b0e96d95b498ba28e0ced956a195b03d6997..48cbd0a5dec09c5557a0c623ace897ded7bc5502